DO-144 Monitoring of docker containers

Durf1985 opened the pull request:
## In the process done:
* Added container monitoring system (cAdvisor, Telegraf)
* Added visualization of metrics and their automatic launch of their tracking (Grafana)
* Added error notification system (Alertmanager)

## How to setup the project:

```
cd your_repo/monitoring/in_all_sub_directory
docker build -t your_docker_username/image_type:image_tag in_all_sub_directory
docker push all_created_image
cd your_repo/docker
docker-compose up -d 
docker compose -f docker-compose-monitoring.yml up -d
```

## How to check the project:

* Follow the link <http://your_docker_machine_ip:3000>, use login and password, which you setup in .env file


## PR checklist

The label with the homework number, in label description - homework lecture and topic

Durf1985 committed:
DO-144 Monitoring of docker containers

Split the docker-compose.yml file into 2 separate files:
* In docker-compose.yml contains the core reddit app services
* In docker-compose-monitoring.yml rendered services that are responsible
  for monitoring.

Added firewall permission on port 8080 in terraform configuration

Added cadvisor to prometheus config

Durf1985 committed:
DO-145 Metrics visualisation using grafana

Grafana config added to docker-compose-monitoring.yml file
Added firewall rule for port 3000
Created dashboard configuration file for grafana

loktionovam commented on docker/docker-compose.yml:
```diff
@@ -1,55 +1,6 @@
 version: '3.3'
 services:
```
I think here should be placed the main part of docker compose - your services (ui, post and so on)

Durf1985 committed:
DO-147-149 UI/Business Monitoring

Added 2 custom dashboards: UI_Service_Monitoring.json, Business_Logic_Monitoring.json
Added parts for alertmanager, there are problems with sending notifications
to slack

Durf1985 committed:
DO-150 Using alerting, setup alermanager

Fixed issues with sending notifications due to an outdated version
alertmanager

loktionovam commented on docker/docker-compose-monitoring.yml:
```diff
@@ -0,0 +1,108 @@
+      - GF_SECURITY_ADMIN_USER=admin
+      - GF_SECURITY_ADMIN_PASSWORD=secret
+    depends_on:
+      - prometheus
```
It is better to not use depends_on especially in microservices (because stopping/starting of services can be really painfull and lead to unpredictable restarts and hang ups). Use `depends_on` only where it is strictly necessary, i.e. your service doesn't support reconnection to database

Durf1985 replied:
Ok, I will delete it in next commit, but it is not my code, this code is from practice

loktionovam commented on docker/docker-compose-monitoring.yml:
```diff
@@ -0,0 +1,108 @@
+      - grafana_data:/var/lib/grafana
+    environment:
+      - GF_SECURITY_ADMIN_USER=admin
+      - GF_SECURITY_ADMIN_PASSWORD=secret
```
This is insecure. Create a variable and store the password there rather than hardcoding it to the docker-compose file

loktionovam commented on docker/docker-compose-monitoring.yml:
```diff
@@ -0,0 +1,108 @@
+  prometheus:
+    image: ${DOCKER_HUB_USERNAME}/prometheus:2.37.6
+    ports:
+      - 9090:9090
```
FYI: publishing prometheus ports (and other exporters) without authentication is good for testing purposes and bad for production usage. Usually you should place your prometheus behind grafana and publish grafana web interface only. Also make sense do not publish your exporters (mongodb, cadvisor and so on) to the Internet too

loktionovam commented on monitoring/alertmanager/config.yml:
```diff
@@ -0,0 +1,11 @@
+global:
+  slack_api_url: 'https://hooks.slack.com/services/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'
```
This is a secret info and should be removed from the git repo. Place this in gitlab CI/CD variables and use a docker entrypoint to template your alertmanager config

Durf1985 committed:
DO-150 Setup alertmanager

Added a Dockerfile to create an image with Alertmanager, since the prom/alertmanager
image uses alpine-scratch, which makes it impossible to install Consul and Consul-template.
Redesigned the Dockerfile to create the final image and added templating
for Slack notification webhooks

Durf1985 committed:
DO-151 Using telegraf to collect docker metrics

Added telegraf dashboard to grafana
Added telegraf Dockerfile and telegraf configuration file
Added telegraf configuration to prometheus config file and
docker-compose-monitoring.yml file

Durf1985 committed:
DO-152 Datasources and dashboards as files

Added preconfigure setting for grafana
